File: lib/profile/data/username_rules.dart
/// Client mirror of `functions/identity/username_rules.js`.
///
/// The SERVER is authoritative for uniqueness. Everything here is either
/// (a) input validation, so the user gets an instant, honest message instead of
/// a round-trip, or (b) an optional availability hint. Neither is ever allowed
/// to decide that a name is free — only the callable's transaction can do that,
/// and it fails closed.
///
/// ── One deliberate difference from the server ───────────────────────────────
/// The server normalises with trim → Unicode NFKC → lowercase. Dart has no
/// built-in NFKC and this app does not carry a Unicode normalisation package,
/// so [normalizeUsername] here is trim → lowercase. NFKC is the identity on
/// ASCII, which is every username the app has ever written, so the two agree in
/// practice. For a hypothetical non-ASCII name they could disagree — and that
/// is safe precisely because this function never authorises anything: a wrong
/// local hint is corrected by the server, which refuses the name.
library;

import 'dart:convert';

import 'package:crypto/crypto.dart';

/// The app's long-standing rule: 3–22 characters, no whitespace.
final RegExp kUsernamePattern = RegExp(r'^\S{3,22}$');

const int kUsernameMinLength = 3;
const int kUsernameMaxLength = 22;

/// Why a username was rejected.
enum UsernameProblem { empty, invalidCharacters, invalidFormat }

/// Normalised, case-insensitive form. See the library note on NFKC.
String normalizeUsername(String? raw) {
  if (raw == null) return '';
  return raw.trim().toLowerCase();
}

/// Display form: trimmed, casing preserved.
String displayUsername(String? raw) => raw == null ? '' : raw.trim();

/// Returns null when the name is acceptable.
UsernameProblem? validateUsername(String? raw) {
  final String display = displayUsername(raw);
  if (display.isEmpty) return UsernameProblem.empty;
  for (final int unit in display.codeUnits) {
    if (unit <= 0x1f || unit == 0x7f) return UsernameProblem.invalidCharacters;
  }
  if (!kUsernamePattern.hasMatch(display)) return UsernameProblem.invalidFormat;
  return null;
}

/// User-facing text for a validation problem.
String usernameProblemMessage(UsernameProblem problem) {
  switch (problem) {
    case UsernameProblem.empty:
      return 'Enter a username.';
    case UsernameProblem.invalidCharacters:
      return 'That username contains characters we can’t use.';
    case UsernameProblem.invalidFormat:
      return 'Usernames are $kUsernameMinLength–$kUsernameMaxLength characters '
          'with no spaces.';
  }
}

/// Path-safe reservation document id, matching the server's derivation.
String usernameIndexKey(String normalized) =>
    sha256.convert(utf8.encode(normalized)).toString();

/// Shown when a name is definitively taken. Matches the signup flow's wording
/// so a collision reads the same wherever it is detected.
const String kUsernameTakenMessage =
    "That username's already taken — try another.";

/// Shown when uniqueness could not be confirmed. Deliberately does NOT claim
/// the name is taken.
const String kUsernameCheckFailedMessage =
    "We couldn't check that username right now. Check your connection and try again.";

/// Shown when the device is offline.
///
/// Uniqueness needs a transaction, and Firestore transactions — unlike batched
/// writes — cannot queue offline. There is no honest way to accept a username
/// change without a connection, so the UI says so rather than pretending.
const String kUsernameNeedsConnectionMessage =
    'Changing your username needs a connection — we have to check nobody else '
    'has taken it.';

File: functions/identity/username_rules.js
// Pure username normalisation, validation and index-key derivation.
//
// No Firebase imports, so the exact rules the callable enforces are the rules
// the unit tests and the backfill migration run. lib/profile/data/username_rules.dart
// is the pinned Dart mirror.
//
// ── Normalisation ───────────────────────────────────────────────────────────
// trim → Unicode NFKC → lowercase. NFKC is the identity on ASCII, so every
// username already written by the app normalises to exactly the value the old
// `username.toLowerCase()` code produced and the backfill is non-destructive.
//
// ── Index key ───────────────────────────────────────────────────────────────
// The reservation document id is sha256(normalised) in hex, NOT the name
// itself. A username may legally contain '/', '.', '..' or a 1500-byte run of
// astral characters, all of which are illegal or dangerous as Firestore
// document ids. Hashing makes the key a fixed-length, path-safe constant.

'use strict';

const crypto = require('crypto');

/** Matches the app's long-standing rule: 3–22 characters, no whitespace. */
const USERNAME_PATTERN = /^\S{3,22}$/u;

const MIN_LENGTH = 3;
const MAX_LENGTH = 22;

/** trim → NFKC → lowercase. Returns '' for anything that is not a string. */
function normalizeUsername(raw) {
  if (typeof raw !== 'string') return '';
  return raw.trim().normalize('NFKC').toLowerCase();
}

/** The display form: trimmed, NFKC, original casing preserved. */
function displayUsername(raw) {
  if (typeof raw !== 'string') return '';
  return raw.trim().normalize('NFKC');
}

/**
 * Validation result: { ok: true } or { ok: false, code, message }.
 * Runs against the DISPLAY form so a name that is only whitespace, or that
 * carries control characters, is rejected before it can reach the index.
 */
function validateUsername(raw) {
  const display = displayUsername(raw);
  if (!display) {
    return { ok: false, code: 'empty', message: 'Enter a username.' };
  }
  // eslint-disable-next-line no-control-regex
  if (/[\u0000-\u001f\u007f]/.test(display)) {
    return {
      ok: false,
      code: 'invalid-characters',
      message: 'That username contains characters we can’t use.',
    };
  }
  if (!USERNAME_PATTERN.test(display)) {
    return {
      ok: false,
      code: 'invalid-format',
      message: `Usernames are ${MIN_LENGTH}–${MAX_LENGTH} characters with no spaces.`,
    };
  }
  return { ok: true };
}

/** Path-safe reservation document id for a normalised username. */
function usernameIndexKey(normalized) {
  return crypto.createHash('sha256').update(String(normalized), 'utf8').digest('hex');
}

/** Convenience: index key straight from a raw, unnormalised name. */
function indexKeyForRaw(raw) {
  return usernameIndexKey(normalizeUsername(raw));
}

module.exports = {
  USERNAME_PATTERN,
  MIN_LENGTH,
  MAX_LENGTH,
  normalizeUsername,
  displayUsername,
  validateUsername,
  usernameIndexKey,
  indexKeyForRaw,
};
